README.md
on-demand

# NAME

on::demand - postpone loading a perl module until it's actually used

# USAGE

    use on::demand "My::Module";
    # the module is not required at this point, speeding up the startup

    my $new = My::Module->new;
    # My::Module is loaded now and real new() method is called

    no on::demand;
    # Force loading of all postponed modules, in alphabetical order

# DESCRIPTION

In a big project with large dependency footprint
loading modules may take quite a while.
This is not a problem for daemons or long-running apps.
However, time saving for simple command-line tools or test scripts
may be significant.

# CONTENTS OF THIS PACKAGE

* `lib` - libraries & modules

* `t` - tests

* `xt` - author tests (not required for installation)

* `README`/`README.md` - this file

# INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

# SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc on::demand

You can also look for information at:

    Github:
        https://github.com/dallaylaen/on-demand-perl

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=on-demand

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/on-demand

    CPAN Ratings
        http://cpanratings.perl.org/d/on-demand

    Search CPAN
        http://search.cpan.org/dist/on-demand/


# LICENSE AND COPYRIGHT

Copyright (C) 2019 Konstantin S. Uvarin

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Module-Lazy

0.02    Sun Mar 3 2019
        - Fix bug with DESTROY being called before loading a package

0.01    Sat Mar 2 2019
        - Initial release. Has `use Module::Load "Some::Module";`
          and `no Module::Load;`


t/lib/on/demand/_/test/sample.pm
package on::demand::_::test::sample;

use strict;
use warnings;
our $VERSION = 42;

our $alive;

sub new {
    $alive++;
    bless {}, shift;
};

sub DESTROY {
    $alive--;
};

1;


t/lib/on/demand/_/test/subclass.pm
package on::demand::_::test::subclass;

use strict;
use warnings;

# Well... No on::demand parents in production please...
# But it works (mostly accidentally) because
# AUTOLOAD uses can() for resolving the real method
# and will stumble upon overloaded can() in parent class
# which will load the parent and call can() on it
# loading the previous class and so on until the full hierarchy is loaded.
use on::demand "on::demand::_::test::sample";

our @ISA = qw( on::demand::_::test::sample );
our $VERSION = 3.14;

sub frobnicate {
    return 42;
};

1;


Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'on::demand',
    AUTHOR           => q{Konstantin S. Uvarin <khedin@cpan.org>},
    VERSION_FROM     => 'lib/on/demand.pm',
    ABSTRACT_FROM    => 'lib/on/demand.pm',
    LICENSE          => 'artistic_2',
    PL_FILES         => {},
    MIN_PERL_VERSION => '5.008',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    BUILD_REQUIRES => {
        'Test::More'      => 0.98, # done_testing + subtest
        'Test::Exception' => 0.03,
    },
    PREREQ_PM => {
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'on-demand*' },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
            bugtracker => {
                web => 'https://github.com/dallaylaen/on-demand-perl/issues',
            },
            repository => {
                type => 'git',
                url => 'https://github.com/dallaylaen/on-demand-perl.git',
                web => 'https://github.com/dallaylaen/on-demand-perl',
            },
        },
        x_contributors => [
            # Alphabetically, with a comma at the end of line
        ],
    },
);


lib/Module/Lazy.pm
package Module::Lazy;

use 5.008;
use strict;
use warnings;
our $VERSION = '0.03';

=head1 NAME

Module::Lazy - postpone loading a module until it's actually used

=head1 SYNOPSIS

    use Module::Lazy "My::Module";
    # My::Module has not been loaded

    my $var = My::Module->new;
    # My::Module is loaded now, and new() method is called

    no Module::Lazy;
    # Force loading of all postponed modules

=head1 DESCRIPTION

In large projects loading all the dependencies may take a lot of time.
This module attempts to reduce the startup time by postponing initialization.
The improvement be significant for unit test scripts
and small command-line tools
which do not utilize all the functionality at once.

This comes at a cost of reduced stability,
as load-time errors are also postponed.
The C<no Module::Lazy> directive is provided to mitigate the risk
by forcing the pending modules to load.

=head1 EXPORTED FUNCTIONS

None.

=head1 METHODS

=cut

use Carp;
use constant DEBUG => !!$ENV{PERL_LAZYLOAD_DEBUG};

=head2 import

When C<use Module::Lazy "Some::Module";> is called,
the module in question is not loaded.
A stub package with the same name is created instead.

Should any method call be performed on the stub package,
it loads the original one and jumps to respective method.

In particular, C<can()> and C<isa()> are overloaded
and will trigger module loading.

Upon loading, C<import> is not called on the target package.
This MAY change in the future.

No extra options (except from target module name) are allowed.

=cut

my $dont = !!$ENV{PERL_LAZYLOAD_DISABLE};
my %seen;
my $inc_stub = "pending load by ".__PACKAGE__;

sub import {
    my ($class, $target, @rest) = @_;

    # bare use statement is ok
    return unless defined $target;

    croak "Usage: use Module::Lazy 'Module::Name'; extra options not supported"
        unless @rest == 0;

    # return ASAP if already loaded by us or Perl itself
    return if $seen{$target};
    my $mod = $target;
    $mod =~ s,::,/,g;
    $mod .= ".pm";

    return if $INC{$mod};

    carp __PACKAGE__.": request to load $target "
        .($seen{$target} ? '(seen)' : '(first time)')
            if DEBUG;
    return _load( $target, $mod )
        if $dont;

    croak "Bad module name '$target'"
        unless $target =~ /^[A-Za-z_][A-Za-z_0-9]*(?:::[A-Za-z_0-9]+)*$/;

    $seen{$target} = $mod;

    # If $target is later require'd directly,
    # autoload and destroy will be overwritten and will cause a warning.
    # Preventing them from being loaded seems like a lesser evil.
    $INC{$mod} = $inc_stub;

    _set_symbol( $target, AUTOLOAD => sub {
        our $AUTOLOAD;
        $AUTOLOAD =~ s/.*:://;
        my $jump = _jump( $target, $AUTOLOAD );
        goto $jump;
    } );

    # Provide DESTROY just in case someone blesses an object directly
    #     without ever loading a module
    _set_symbol( $target, DESTROY => _jump( $target, DESTROY => "no_die" ) );

    # If somebody calls Module->can("foo"), we can't really tell
    # without loading, so override it
    foreach (qw( can isa )) {
        _set_symbol( $target, $_ => _jump( $target, $_ ) );
    };

    # Provide a fake version for `use My::Module 100.500`
    _set_symbol( $target, VERSION => 10**9 );
};

=head2 unimport

Calling C<no Module::Lazy;> or, alternatively, C<Module::Lazy-E<gt>unimport;>
will cause all postponed modules to be loaded immediately,
in alphabetical order.

This may be useful to avoid deferred errors and/or side effects
of module loading.

No extra options to unimport are supported.

=cut

sub unimport {
    my $class = shift;

    croak "usage: no Module::Lazy;"
        if @_;

    carp __PACKAGE__.": unimport called"
        if DEBUG;

    $dont++;
    # sort keys to ensure load order stability in case of bugs
    foreach (sort keys %seen) {
        # some modules may have been already loaded, skip if so
        _inflate($_) if $seen{$_};
    };
};

my %cleanup_symbol;
sub _inflate {
    my $target = shift;

    # TODO distinguish between "not seen" and "already loaded"
    my $mod = delete $seen{$target};
    croak "Module '$target' is unknown Module::Lazy, or already loaded. Please file a bug"
        unless $mod;

    carp "Module '$target' wasn't preloaded by Module::Lazy. Please file a bug"
        unless $INC{$mod};

    return carp "Module '$target' already loaded elsewhere from '$INC{$mod}'"
        unless $INC{$mod} eq $inc_stub;

    # reset stub methods prior to loading
    foreach (keys %{ $cleanup_symbol{$target} || {} }) {
        _unset_symbol( $target, $_ );
    };

    # reset fake $VERSION
    _set_symbol( $target, VERSION => undef );

    # make the module loadable again
    delete $INC{$mod};
    _load( $target, $mod );
};

sub _load {
    my ($target, $mod) = @_;

    carp __PACKAGE__.": loading $target from $mod"
        if DEBUG;

    package
        Module::Lazy::_::quarantine;

    local $Carp::Internal{ __PACKAGE__ } = 1;
    require $mod;
    # TODO maybe $target->import()
};

sub _jump {
    my ($target, $todo, $nodie) = @_;

    return sub {
        _inflate( $target );

        my $jump = $target->can($todo);
        goto $jump
            if $jump; # TODO should also check it's a CODEREF

        croak qq{Can't locate object method "$todo" via package "$target"}
            unless $nodie;
    };
};

sub _set_symbol {
    my ($target, $name, $ref) = @_;

    if (ref $ref) {
        # really update symbol table
        $cleanup_symbol{$target}{$name}++;
        no strict 'refs'; ## no critic
        *{ $target."::".$name } = $ref;
    } else {
        # just set scalar
        no strict 'refs'; ## no critic
        ${ $target.'::'.$name } = $ref;
    };
};

sub _unset_symbol {
    my ($target, $name) = @_;

    no strict 'refs'; ## no critic
    # because package scalars are _special_,
    # move SCALAR ref around the destruction
    # just in case someone referenced it before module was loaded
    my $save = \${ $target."::".$name };
    delete ${ $target."::" }{ $name };
    *{ $target.'::'.$name } = $save;
};

=head1 ENVIRONMENT

If C<PERL_LAZYLOAD_DEBUG> is set and true,
warns about module loading via Carp.

If C<PERL_LAZYLOAD_DISABLE> is set and true,
don't try to lazyload anything - just go straight to C<require>.

(That's roughly equivalent to C<perl -M-Module::Lazy> on command line).

=head1 CAVEATS

=over

=item * The following symbols are currently replaced by stubs
in the module to be loaded: C<AUTOLOAD>, C<DESTROY>, C<can>, C<isa>.

=item * If a module was ever lazyloaded, a normal C<require> would do nothing.
A method must be called to inflate the module.

This is done so because a normal require would partially overwrite
stub methods and potentially wreak havoc.

=item * A fake $VERSION = 10**9 is generated so that C<use Module x.yy>
doesn't die. This value is erased before actually loading the module.

=back



=head1 BUGS

=over

=item * C<use mro 'c3';> does not work with lazy-loaded parent classes.

=item * C<import()> is not called on the modules being loaded.
The decision is yet to be made whether it's good or bad.

=item * no way to preload prototyped exported functions
(that's what L<autouse> does),
but maybe there should be?

=item * certainly not enough interoperability tests.

=back

Please report bugs via github or RT:

=over

=item * L<https://github.com/dallaylaen/module-lazy-perl/issues>

=item * C<bug-module-lazy at rt.cpan.org>

=item * L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Module-Lazy>

=back

=head1 SUPPORT

You can find documentation for this module with the C<perldoc> command.

    perldoc Module::Lazy

You can also look for information at:

=over 4

=item * github: L<https://github.com/dallaylaen/module-lazy-perl>

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Lazy>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Module-Lazy>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Module-Lazy>

=item * Search CPAN

L<http://metacpan.org/pod/Module::Lazy/>

=back

=head1 SEE ALSO

L<autouse> is another module with similar idea, however,
it does it for imported functions rather than methods.

=head1 ACKNOWLEDGEMENTS

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Konstantin S. Uvarin, C<< <khedin@cpan.org> >>

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Module::Lazy


lib/on/demand.pm
package on::demand;

use 5.008;
use strict;
use warnings;
our $VERSION = '0.01';

=head1 NAME

on::demand - postpone loading a module until it's actually used

=head1 SYNOPSIS

    use on::demand "My::Module";
    # My::Module has not been loaded

    my $var = My::Module->new;
    # My::Module is loaded now, and new() method is called

    no on::demand;
    # Force loading of all postponed modules

=head1 DESCRIPTION

In large projects loading all the dependencies may take a lot of time.
This module attempts to reduce the startup time by postponing initialization.
The improvement be significant for unit test scripts
and small command-line tools
which do not utilize all the functionality at once.

This comes at a cost of reduced stability,
as load-time errors are also postponed.
The C<no on::demand> directive is provided to mitigate the risk
by forcing the pending modules to load.

=head1 EXPORTED FUNCTIONS

None.

=head1 METHODS

=cut

use Carp;

=head2 import

When C<use on::demand "Some::Module";> is called,
the module in question is not loaded.
A stub package with the same name is created instead.

Should any method call be performed on the stub package,
it loads the original one and jumps to respective method.

In particular, C<can()> and C<isa()> are overloaded
and will trigger module loading.

Upon loading, C<import> is not called on the target package.
This MAY change in the future.

No extra options (except from target module name) are allowed.

=cut

my %seen;
sub import {
    my ($class, $target, @rest) = @_;

    croak "Usage: use on::demand 'Module::Name';"
        unless defined $target and @rest == 0;

    # return ASAP if already loaded by us or Perl itself
    return if $seen{$target};
    my $mod = $target;
    $mod =~ s,::,/,g;
    $mod .= ".pm";
    return if $INC{$mod};

    croak "Bad module name '$target'"
        unless $target =~ /^[A-Za-z_][A-Za-z_0-9]*(?:::[A-Za-z_0-9]+)*$/;

    $seen{$target} = $mod;

    our $AUTOLOAD;
    _set_function( $target, AUTOLOAD => sub {
        _load( $target );
        my $jump = _jump( $target, $AUTOLOAD );
        goto $jump;
    } );

    foreach (qw( can isa )) {
        my $name = $_; # separate variable to close over
        _set_function( $target, $name => sub {
            _load( $target );
            my $jump = _jump( $target, $name );
            goto $jump;
        });
    };
};

=head2 unimport

Calling C<no on::demand;> or, alternatively, C<on::demand-E<gt>unimport;>
will cause all postponed modules to be loaded immediately,
in alphabetical order.

This may be useful to avoid deferred errors and/or side effects
of module loading.

No extra options to unimport are supported.

=cut

sub unimport {
    my $class = shift;

    croak "usage: no on::demand;"
        if @_;

    # sort keys to ensure load order stability in case of bugs
    foreach (sort keys %seen) {
        _load($_);
    };
};

my %known_method;
sub _load {
    my $target = shift;

    my $mod = delete $seen{$target};
    croak "Module '$target' was never loaded via on::demand, that's possibly a bug"
        unless $mod;

    # reset stub methods prior to loading
    foreach (keys %{ $known_method{$target} || {} }) {
        _set_function( $target, $_ => undef );
    };

    package
        on::demand::_::quarantine;

    local $Carp::Internal{ __PACKAGE__ } = 1;
    require $mod;
    # TODO maybe import()
};

sub _jump {
    my ($target, $todo, $nodie) = @_;

    $todo =~ s/.*:://;
    my $jump = $target->can($todo);

    croak qq{Can't locate object method "$todo" via package "$target"}
        unless $jump or $nodie;

    return $jump;
};

sub _set_function {
    my ($target, $name, $code) = @_;

    if (ref $code) {
        $known_method{$target}{$name}++;
        no strict 'refs'; ## no critic
        *{ $target."::".$name } = $code;
    } else {
        no strict 'refs'; ## no critic
        delete ${ $target."::" }{ $name };
    };
};

=head1 AUTHOR

Konstantin S. Uvarin, C<< <khedin@cpan.org> >>

=head1 BUGS

=over

=item * import() is not called on the modules being loaded.
The decision is yet to be made whether it's good or bad.

=item * C<on on::demand> should prevent further demand loading.

=item * no way to preload prototyped exported functions
(that's what L<autouse> does),
but maybe there should be?

=item * certainly not enough interoperability tests (C<use mro 'c3'>?).

=back

Please report bugs via github or RT:

=over

=item * L<https://github.com/dallaylaen/on-demand-perl/issues>

=item * C<bug-assert-refute-t-deep at rt.cpan.org>

=item * L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=on-demand>

=back

=head1 SUPPORT

You can find documentation for this module with the C<perldoc> command.

    perldoc on::demand

You can also look for information at:

=over 4

=item * github: L<https://github.com/dallaylaen/on-demand-perl>

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=on-demand>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/on-demand>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/on-demand>

=item * Search CPAN

L<http://metacpan.org/pod/on::demand/>

=back

=head1 SEE ALSO

L<autouse> is another module with similar idea, however,
it does it for imported functions rather than methods.

=head1 ACKNOWLEDGEMENTS

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Konstantin S. Uvarin.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of on::demand
